plugin/core/sessions.py

```diff
@@ -1898,10 +1910,11 @@ def deduce_payload(
                 return res
         elif "id" in payload:
             response_id = int(payload["id"])
-            handler, result, is_error = self.response_handler(response_id, payload)
-            response_tuple = (handler, result, None, None, None)
-            self._logger.incoming_response(response_id, result, is_error)
-            return response_tuple
+            handler, method, result, is_error = self.response_handler(response_id, payload)
+            self._logger.incoming_response(response_id, deepcopy(result), is_error)
```

Comment:
Is it necessary to deepcopy? It can be a significant performance penalty since when thinking about typescript returning potentially megabytes of completions.

Reply:
If the result is a dict, and you use normal `copy` and later modify `result["somekey"]`, then the copy will also see the new value. Therefore we need to use deepcopy here.

I guess the only alternative would be to not copy at all and do the logging on the main thread instead of async, but that would probably have an even bigger performance impact.

Reply:
Actually I'm not sure now. When I use `from copy import copy` and modify `deepcopy(result)` to `copy(result)` it does not work, i.e. the modified payload gets logged into the console. But when I run the following in the Python REPL, it does work:
```
>>> from copy import copy
>>> old = {"x": "foo"}
>>> copied = copy(old)
>>> old["x"] = "bar"
>>> copied
{'x': 'foo'}
```

Strange

Reply:
Another idea would be to gate that whole line behind a `if userprefs().log_server`, so it would only impact performance if logging is enabled. But from the code I can only see the PanelLogger to honor this setting inside its logging methods, and idk how the remote logger works. Would it break anything to not call this method at this place here, if the setting is disabled?

Reply:
I always run with logging enabled so I wouldn't be happy with performance being affected in that case.

I guess the question can also be asked whether we have to log the value as received from the server or it's OK to log it after object being potentially modified. I guess either could cause confusion, depending on what one is looking for.

Reply:
Basically if I don't see an `from copy import deepcopy` line anymore I'll merge this :)

Reply:
deepcopy is still there.

Reply:
So `on_server_response_async`, or maybe `on_receive_response_async` to be closer to the existing naming scheme? ("pre" isn't really applicable here I guess)

Reply:
It is indeed on the async thread.

But why is there a `set_timeout_async` then in
https://github.com/sublimelsp/LSP/blob/418d9931e62e63d0484279995b79843dbdfc051d/plugin/core/windows.py#L505-L517

... when all the logging methods and therefore `PanelLogger.log()` is already called from the async thread anyways? This seems to be the reason why a following modification to the payload would currently end up in the log panel too. Shouldn't the logging work as expected, (i.e. original payload), and without the need for deepcopy if we remove the `set_timeout_async` in `log()`?

Reply:
It’s been a long time since I looked at that part of the code, but I think the set_timeout_async is indeed not necessary.

Reply:
Yes, having `receive` would be more in line though it gets pretty wordy.

Though `on_receive_response_async` doesn't sound right as it could then be misunderstood as client sending the response. So it would have to be `on_receive_server_response_async`.

Let's see how the names could look like if we'd follow the current convention...

## Existing
 - on_pre_send_request_async
 - on_pre_send_notification_async
## This one
 - on_receive_server_response_async
## Potential
 - on_receive_client_response_async
 - on_pre_server_request_async
 - on_pre_client_request_async

(The `*_client_*` ones could still be useful IMO as one might want to let the client use the default logic and then still modify something).

I suppose this would work after all?

Reply:
Well, that's not consistent either as existing APIs don't contain the `*_client_*` part...

Reply:
Proposed simplification below. The current APIs could be deprecated and replaced with new ones eventually (I might be the only one using one of them anyway).

## Existing
 - on_client_notification_async
 - on_client_request_async
## This one
 - on_server_response_async
## Potential
 - on_client_response_async
 - on_server_request_async

Reply:
> It’s been a long time since I looked at that part of the code, but I think the set_timeout_async is indeed not necessary.

I agree